### PR TITLE
[CVS-94911] Fix difference between train and validation normalization pipeline

### PR DESCRIPTION
**Pull request by jaegukhyun**
This PR includes solve issue CVS-94911.
Previous implementation only look into pipeline of the most inside dataset
I change the code to also look into outside dataset
I think we should refactor this function recursive way to handle various dataset settings.

### CODE REVIEW

**Review by goodsong81 — COMMENTED** [review]
Could you replicate this change to feature/otx?
@harimkang Or, do you have plan to move changes to develop to feature/otx?

**Comment by harimkang** [comment]
> Could you replicate this change to feature/otx? @harimkang Or, do you have plan to move changes to develop to feature/otx?

Commits with (& after) tiling are not included in `feature/otx`. However, I know that `develop` currently has some issues. When those issues are resolved, I will reflect `develop` to `feature/otx` once.

**Comment by jaegukhyun** [comment]
> Could you replicate this change to feature/otx?

I can directly change otx/feature branch to reflect this PR. However if there are plan to merge develop and otx/feature branch, then I can wait for it.

**Comment by sungchul1** [comment]
I checked all of `to_rgb` are changed to `False`.
If modified parts are fine, you can merge it.

**Review by sungchul1 — DISMISSED** [review]

**Comment by eugene123tw** [comment]
run ote-test

**Review by eugene123tw — CHANGES_REQUESTED** [review]
I'm getting this error running tests. Could you take a look?

```
Traceback (most recent call last):
  File "/tmp/ote_cli/model-preparation-algorithm/venv/bin/ote", line 33, in <module>
    sys.exit(load_entry_point('ote-cli', 'console_scripts', 'ote')())
  File "/home/yuchunli/git/training_extensions/ote_cli/ote_cli/tools/ote.py", line 67, in main
    globals()[f"ote_{name}"]()
  File "/home/yuchunli/git/training_extensions/ote_cli/ote_cli/tools/train.py", line 175, in main
    task.train(dataset, output_model, train_parameters=TrainParameters())
  File "/home/yuchunli/git/training_extensions/external/model-preparation-algorithm/mpa_tasks/apis/detection/task.py", line 528, in train
    results = self._run_task(stage_module, mode="train", dataset=dataset, parameters=train_parameters)
  File "/home/yuchunli/git/training_extensions/external/model-preparation-algorithm/mpa_tasks/apis/task.py", line 112, in _run_task
    output = workflow.run(
  File "/home/yuchunli/git/training_extensions/external/model-preparation-algorithm/submodule/mpa/det/trainer.py", line 126, in run
    DetectionTrainer.train_worker(
  File "/home/yuchunli/git/training_extensions/external/model-preparation-algorithm/submodule/mpa/det/trainer.py", line 161, in train_worker
    train_detector(
  File "/home/yuchunli/git/training_extensions/external/mmdetection/submodule/mmdet/apis/train.py", line 260, in train_detector
    runner.run(data_loaders, cfg.workflow, compression_ctrl=compression_ctrl)
  File "/tmp/ote_cli/model-preparation-algorithm/venv/lib/python3.8/site-packages/mmcv/runner/epoch_based_runner.py", line 127, in run
    epoch_runner(data_loaders[i], **kwargs)
  File "/home/yuchunli/git/training_extensions/ote_sdk/ote_sdk/utils/argument_checks.py", line 250, in validate
    return function(**input_parameters_values_map)
  File "/home/yuchunli/git/training_extensions/external/model-preparation-algorithm/mpa_tasks/utils/runner.py", line 53, in train
    for i, data_batch in enumerate(self.data_loader):
  File "/tmp/ote_cli/model-preparation-algorithm/venv/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 517, in __next__
    data = self._next_data()
  File "/tmp/ote_cli/model-preparation-algorithm/venv/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 557, in _next_data
    data = self._dataset_fetcher.fetch(index)  # may raise StopIteration
  File "/tmp/ote_cli/model-preparation-algorithm/venv/lib/python3.8/site-packages/torch/utils/data/_utils/fetch.py", line 44, in fetch
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "/tmp/ote_cli/model-preparation-algorithm/venv/lib/python3.8/site-packages/torch/utils/data/_utils/fetch.py", line 44, in <listcomp>
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "/home/yuchunli/git/training_extensions/external/mmdetection/submodule/mmdet/datasets/dataset_wrappers.py", line 475, in __getitem__
    return self.pipeline(self.tile_dataset[idx])
  File "/home/yuchunli/git/training_extensions/external/mmdetection/submodule/mmdet/datasets/pipelines/compose.py", line 40, in __call__
    data = t(data)
  File "/home/yuchunli/git/training_extensions/external/mmdetection/submodule/mmdet/datasets/pipelines/formating.py", line 322, in __call__
    img_meta[key] = results[key]
KeyError: 'ignored_labels'
FAILED
```

File: external/model-preparation-algorithm/mpa_tasks/apis/detection/task.py [hunk 1]
```diff
@@ -344,6 +344,21 @@ def patch_color_conversion(pipeline):
...
+            if pipeline is not None:
+                for pipeline_step in pipeline:
+                    if pipeline_step.type == "LoadImageFromFile":
+                        pipeline_step.type = "LoadImageFromOTEDataset"
+                    if pipeline_step.type == "LoadAnnotations":
+                        pipeline_step.type = "LoadAnnotationFromOTEDataset"
+                        pipeline_step.domain = domain
+                        pipeline_step.min_size = cfg.pop("min_size", -1)
+                    if subset == "train" and pipeline_step.type == "Collect":
+                        pipeline_step = BaseTask._get_meta_keys(pipeline_step)
```
Comment (harimkang) [inline external/model-preparation-algorithm/mpa_tasks/apis/detection/task.py:360]: The `ignored_labels` key error in the comment is a problem that can be solved by going through this `_get_meta_keys` function. I can't seem to get into this line, so please check.

**Comment by harimkang** [comment]
> I'm getting this error running tests. Could you take a look?
> 
> ```
> Traceback (most recent call last):
>   File "/tmp/ote_cli/model-preparation-algorithm/venv/bin/ote", line 33, in <module>
>     sys.exit(load_entry_point('ote-cli', 'console_scripts', 'ote')())
>   File "/home/yuchunli/git/training_extensions/ote_cli/ote_cli/tools/ote.py", line 67, in main
>     globals()[f"ote_{name}"]()
>   File "/home/yuchunli/git/training_extensions/ote_cli/ote_cli/tools/train.py", line 175, in main
>     task.train(dataset, output_model, train_parameters=TrainParameters())
>   File "/home/yuchunli/git/training_extensions/external/model-preparation-algorithm/mpa_tasks/apis/detection/task.py", line 528, in train
>     results = self._run_task(stage_module, mode="train", dataset=dataset, parameters=train_parameters)
>   File "/home/yuchunli/git/training_extensions/external/model-preparation-algorithm/mpa_tasks/apis/task.py", line 112, in _run_task
>     output = workflow.run(
>   File "/home/yuchunli/git/training_extensions/external/model-preparation-algorithm/submodule/mpa/det/trainer.py", line 126, in run
>     DetectionTrainer.train_worker(
>   File "/home/yuchunli/git/training_extensions/external/model-preparation-algorithm/submodule/mpa/det/trainer.py", line 161, in train_worker
>     train_detector(
>   File "/home/yuchunli/git/training_extensions/external/mmdetection/submodule/mmdet/apis/train.py", line 260, in train_detector
>     runner.run(data_loaders, cfg.workflow, compression_ctrl=compression_ctrl)
>   File "/tmp/ote_cli/model-preparation-algorithm/venv/lib/python3.8/site-packages/mmcv/runner/epoch_based_runner.py", line 127, in run
>     epoch_runner(data_loaders[i], **kwargs)
>   File "/home/yuchunli/git/training_extensions/ote_sdk/ote_sdk/utils/argument_checks.py", line 250, in validate
>     return function(**input_parameters_values_map)
>   File "/home/yuchunli/git/training_extensions/external/model-preparation-algorithm/mpa_tasks/utils/runner.py", line 53, in train
>     for i, data_batch in enumerate(self.data_loader):
>   File "/tmp/ote_cli/model-preparation-algorithm/venv/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 517, in __next__
>     data = self._next_data()
>   File "/tmp/ote_cli/model-preparation-algorithm/venv/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 557, in _next_data
>     data = self._dataset_fetcher.fetch(index)  # may raise StopIteration
>   File "/tmp/ote_cli/model-preparation-algorithm/venv/lib/python3.8/site-packages/torch/utils/data/_utils/fetch.py", line 44, in fetch
>     data = [self.dataset[idx] for idx in possibly_batched_index]
>   File "/tmp/ote_cli/model-preparation-algorithm/venv/lib/python3.8/site-packages/torch/utils/data/_utils/fetch.py", line 44, in <listcomp>
>     data = [self.dataset[idx] for idx in possibly_batched_index]
>   File "/home/yuchunli/git/training_extensions/external/mmdetection/submodule/mmdet/datasets/dataset_wrappers.py", line 475, in __getitem__
>     return self.pipeline(self.tile_dataset[idx])
>   File "/home/yuchunli/git/training_extensions/external/mmdetection/submodule/mmdet/datasets/pipelines/compose.py", line 40, in __call__
>     data = t(data)
>   File "/home/yuchunli/git/training_extensions/external/mmdetection/submodule/mmdet/datasets/pipelines/formating.py", line 322, in __call__
>     img_meta[key] = results[key]
> KeyError: 'ignored_labels'
> FAILED
> ```

Hi @eugene123tw, I talked about this with @jaegukhyun . Currently, other datasets in OTX are able to receive ignored_labels as collection inputs, but the tiling dataset seems to be an issue because it is not handled that way. Once it is tiling, Jaeguk will fix it so that the function is not called. It would be nice to be able to receive `ignored_labels` from the tiling dataset in the future. For now, let's fix it as a temporary solution.

**Comment by jaegukhyun** [comment]
run ote-test

**Comment by jaegukhyun** [comment]
run ote-test

**Comment by jaegukhyun** [comment]
run ote-test

**Comment by jaegukhyun** [comment]
run ote-test

**Comment by jaegukhyun** [comment]
http://validationreports.sclab.intel.com:8006/reports/build_number_report?trimReq=on&status=FAIL&test_session_build_number=pr-ote-test-697&environment=iotg-dev-workstation-09
@harimkang, @eugene123tw It's time to merge this PR, could you check this once more?

**Review by eugene123tw — APPROVED** [review]

**Review by harimkang — APPROVED** [review]

**Pull request merged**